README.md
# 私有化 CI/CD 解决方案

[![GitHub stars](https://img.shields.io/github/stars/khs1994-docker/ci.svg?style=social&label=Stars)](https://github.com/khs1994-docker/ci) [![star](https://gitee.com/khs1994-docker/ci/badge/star.svg?theme=dark)](https://gitee.com/khs1994-docker/ci/stargazers)

* [支持文档](docs)

* [问题反馈](https://github.com/khs1994-docker/ci/issues)

* [更多信息](https://www.khs1994.com/categories/CI/Drone/)

## 如何工作

**1.** 本地制作 Docker 镜像推送到私有仓库（Docker Registry）

**2.** 本地开发项目，项目根目录编写 `.drone.yml` 文件，推送到 git （例如，GitHub，Gogs ...）

**3** Drone 自动拉取代码完成编译，部署 (Drone 本质就是在指定的容器中运行指定的命令，通过项目根目录中的 `.drone.yml` 文件指定)。

**4** 支持哪些编程语言？理论上支持所有的编程语言！

## TLS

本教程通过 IP + 不同端口 来提供不同的服务，如果你想要通过域名（`TLS`）来提供不同的服务，请查看 [README.TLS.md](README.TLS.md)。

## 准备

* 有公网 IP 的云服务器（推荐，但不是必须）

* Docker CE v18.06 Stable +

* docker-compose v1.22.0 +

* 知道如何注册 GitHub App (GitHub only)

* `brew install gnu-sed` (macOS only)

## 快速开始

### 安装

```bash
$ git clone https://github.com/khs1994-docker/ci.git ~/ci

$ cd ci
```

### 修改配置

执行以下命令完成初始化，然后修改配置。

```bash
$ ./ci
```

### 配置 hosts

修改 `.env` 中的 `CI_HOST` 变量值为 `你自己的 IP`(例如 `云服务器公网 IP`、`路由器分配给电脑的 IP`)

### MySQL 密码(可选)

修改 `secrets/mysql.env` 中的 `MYSQL_ROOT_PASSWORD` 变量值为 MySQL 密码。

### 启用软件

修改 `.env` 中的 `CI_INCLUDE` 变量。

### 使用外部服务？(高级选项)

编辑 `.env` 文件，编辑 `CI_INCLUDE` 变量，去掉内置的软件名，之后填写外部服务的相关配置

```bash
# CI_INCLUDE="drone-server drone-agent gogs registry mysql redis"

CI_INCLUDE="drone-server drone-agent gogs registry"

CI_EXTERNAL_MYSQL_HOST=
CI_EXTERNAL_MYSQL_PORT=
CI_EXTERNAL_MYSQL_USERNAME=
CI_EXTERNAL_MYSQL_PASSWORD=
CI_EXTERNAL_MYSQL_DATABASE=

CI_EXTERNAL_REDIS_HOST=
```

### 选择 Git 服务商

编辑 `docker-compose.override.yml` 文件最下方

默认使用 `Gogs` ，如需使用 `GitHub` 按如下内容修改

```yaml

...

services:
  drone-server:
    # << : *gogs
    << : *github
    # << : *coding
    # << : *khs1994
```

## 启动

```bash
$ ./ci up [-d] [--reset]
```

## 访问服务

> 能不开放端口尽量不开放（例如数据库、缓存）。

* git HTTP **3000**

* git SSH **8022**

* drone **8000**

* registry **5000**

## 启用构建

在 `Drone` 页面登录账号，在项目列表的右边打开开关，将项目推送到 Git，可以看到 Drone 开始构建项目。

## 使用示例

* [HEXO](https://github.com/khs1994-php/hexo)

* [GitBook](https://github.com/khs1994-php/gitbook)

* [PHP](https://github.com/khs1994-php/tencent-ai)

## 最佳实践

https://www.khs1994.com/categories/CI/Drone/

## More Information

* [Gogs](https://github.com/gogits/gogs)

* [Gogs Docker](https://github.com/gogits/gogs/tree/master/docker)

* [Drone](https://github.com/drone)

* [Drone Documents](http://docs.drone.io/)

* [Drone Docker](https://store.docker.com/profiles/drone)


CHANGELOG.md
* `Docker Compose` `1.22.0` `3.7`

* 合并 compose 文件，通过 `CI-CLI` 控制

* `Docker Compose` `1.20.1` `3.6`

* 删除 `docker-compose.yml`，改为以 git 服务商为后缀的 `Compose` 文件

* `Docker Compose` `1.18.0` `3.5`

* Drone 0.8 配置 coding.net 出现错误，原因不明

* `Docker Compose` `1.17.1` `3.4`


ci.sh
#!/usr/bin/env bash

print_help_info(){
  exec echo "
CI-CLI v18.06

Official WebSite https://ci.khs1994.com

Usage: ./ci.sh COMMAND

Debug: [env] DEBUG=1 ./ci.sh COMMAND options

Options:

  --port-port            开放 MySQL Redis 端口

  -d                     后台运行

  --use-external-nginx   使用外部 NGINX = 后边值为配置文件路径 (TLS Only)

Commands:

  change-mysql-root-password

  up           [-d]   [--open-port]

  config              [--open-port]

  up-tls       [-d]   [--open-port]  [--use-external-nginx=/etc/nginx/conf.d]

  tls-config          [--open-port]

  down

  swarm-deploy

  swarm-remove

  k8s-deploy

  k8s-remove

  reset           恢复原始状态

Read './docs/*.md' for more information about CLI commands.

You can open issue in [ https://github.com/khs1994-docker/ci/issues ] when you meet problems.

You must Update .env file when update this project.

Donate https://zan.khs1994.com
"
}

_cp(){
  if [ ! -f $1 ];then cp $2 $1; fi
}

_init(){
  _cp .env .env.example
  _cp docker-compose.override.yml docker-compose.override.demo.yml

  cd config

  _cp gogs/app.ini gogs/app.example.ini
  _cp registry/config.yml registry/config.example.yml

  cd ../webhooks

  _cp .env .env.example
  _cp update.js update.example.js
  _cp update.sh update.example.sh

  cd ../

  cd config/nginx

  _cp docker-registry.conf demo-docker-registry.config
  _cp drone.conf demo-drone.config
  _cp gogs.conf demo-gogs.config

  cd ../../

  command -v docker-compose > /dev/null 2>&1

  if [ $? -ne 0 ];then exec echo "docker-compose not install" ; fi
}

_reset_(){
  rm -rf config/gogs/app.ini \
      config/registry/config.yml \
      webhooks/update.js \
      webhooks/update.sh \
      config/nginx/*.conf
}

_reset(){
  _reset_
  rm -rf .env webhooks/.env docker-compose.override.yml
}

_change_mysql_root_password(){
 echo 1
}

_sed_common(){
  cd config
  sed -i "s#{{ DB_TYPE }}#${CI_DB_TYPE:-mysql}#g" gogs/app.ini
  sed -i "s#{{ DB_HOST }}#${CI_EXTERNAL_MYSQL_HOST:-mysql}:${CI_EXTERNAL_MYSQL_PORT:-3306}#g" gogs/app.ini
  sed -i "s#{{ DB_DATABASE }}#${CI_EXTERNAL_MYSQL_DATABASE:-$MYSQL_DATABASE}#g" gogs/app.ini
  sed -i "s#{{ DB_USERNAME }}#${CI_EXTERNAL_MYSQL_USERNAME:-root}#g" gogs/app.ini
  sed -i "s#{{ DB_PASSWORD }}#${CI_EXTERNAL_MYSQL_PASSWORD:-$MYSQL_ROOT_PASSWORD}#g" gogs/app.ini

  sed -i "s#{{ SSH_PORT }}#${CI_GOGS_SSH_PORT:-8022}#g" gogs/app.ini

  sed -i "s#{{ MAIL_HOST }}#${CI_MAIL_HOST}#g" gogs/app.ini
  sed -i "s#{{ MAIL_FROM }}#${CI_MAIL_FROM}#g" gogs/app.ini
  sed -i "s#{{ MAIL_USERNAME }}#${CI_MAIL_USERNAME}#g" gogs/app.ini
  sed -i "s#{{ MAIL_PASSWORD }}#${CI_MAIL_PASSWORD}#g" gogs/app.ini
  cd -
}

_up(){
    _reset_ ; _init
    _sed_common
    cd config
    sed -i "s#{{ CI_DOMAIN }}#${CI_HOST:-192.168.199.100}#g" gogs/app.ini

    sed -i "s#{{ CI_DOMAIN_FULL }}#${CI_HOST:-192.168.199.100}#g" gogs/app.ini

    sed -i "s#{{ CI_GOGS_PORT }}#${CI_GOGS_PORT:-3000}#g" gogs/app.ini

    sed -i "s#{{ PROTOCOL }}#http#g" gogs/app.ini
    sed -i "s!^CERT_FILE.*!#CERT_FILE!g" gogs/app.ini
    sed -i "s!^KEY_FILE.*!#KEY_FILE!g" gogs/app.ini
    sed -i "s!^TLS_MIN_VERSION.*!#TLS_MIN_VERSION!g" gogs/app.ini

    sed -i "s#{{ REDIS_HOST }}#${CI_EXTERNAL_REDIS_HOST:-$REDIS_HOST}#g" registry/config.yml
    sed -i "s#{{ WEBHOOKS_HOST }}#${WEBHOOKS_HOST:-http://192.168.199.100}#g" registry/config.yml
    cd -
    sed -i "s#{{ DRONE_HOST }}#http://${CI_HOST:-192.168.199.100}:${CI_DRONE_PORT:-8000}#g" docker-compose.override.yml
    sed -i "s#{{ DRONE_GOGS_URL }}#http://${CI_HOST:-192.168.199.100}:${CI_GOGS_PORT:-3000}#g" docker-compose.override.yml

    CI_INCLUDE=${CI_INCLUDE:-drone-server drone-agent gogs registry}

   # based posrt don't up nginx

    for soft in $CI_INCLUDE
    do
      if [ $soft = 'nginx' ];then
        continue
      fi
      CI_INCLUDE_TARGET+="$soft "
    done

    docker-compose ${COMPOSE_FILE:-} up ${opt:-} ${CI_INCLUDE_TARGET}
}

_config(){
  docker-compose ${COMPOSE_FILE:-} config
}

_up-tls(){
    _reset_ ; _init
    _sed_common
    cd config
    sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" gogs/app.ini

    sed -i "s#{{ CI_DOMAIN_FULL }}#git.${CI_DOMAIN:-t.khs1994.com}#g" gogs/app.ini

    sed -i "s#{{ CI_GOGS_PORT }}#${CI_GOGS_PORT:-443}#g" gogs/app.ini

    sed -i "s#{{ PROTOCOL }}#https#g" gogs/app.ini

    sed -i "s#{{ REDIS_HOST }}#${CI_EXTERNAL_REDIS_HOST:-$REDIS_HOST}#g" registry/config.yml
    sed -i "s#{{ WEBHOOKS_HOST }}#${WEBHOOKS_HOST:-https://ci.t.khs1994.com/docker/webhooks}#g" registry/config.yml
    cd -
    sed -i "s#{{ DRONE_HOST }}#https://drone.${CI_DOMAIN:-t.khs1994.com}#g" docker-compose.override.yml
    sed -i "s#{{ DRONE_GOGS_URL }}#https://git.${CI_DOMAIN:-t.khs1994.com}#g" docker-compose.override.yml

    sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" config/nginx/docker-registry.conf
    sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" config/nginx/drone.conf
    sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" config/nginx/gogs.conf

    _sed_external_nginx(){
      # 使用外部 NGINX
      cd config/nginx
      for file in $( ls *.conf )
      do
        sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" $file
        sed -i "s#{{ REGISTRY_UPSTREAM }}#${CI_HOST:-192.168.199.100}#g" $file
        sed -i "s#{{ DRONE_UPSTREAM }}#${CI_HOST:-192.168.199.100}#g" $file
        sed -i "s#{{ GOGS_UPSTREAM }}#${CI_HOST:-192.168.199.100}#g" $file
      done
        cp -a $PWD/*.conf $NGINX_CONF
        mkdir -p $NGINX_CONF/ssl || echo
        if ! [ -f ssl/$CI_DOMAIN.crt ];then
            cp -a $PWD/ssl/*  $NGINX_CONF/ssl
        fi
      cd -
    }

    _sed_nginx(){
      # 使用内部 NGINX
      cd config/nginx
      for file in `ls *.conf`
      do
        sed -i "s#{{ CI_DOMAIN }}#${CI_DOMAIN:-t.khs1994.com}#g" $file
        sed -i "s#{{ REGISTRY_UPSTREAM }}#registry#g" $file
        sed -i "s#{{ DRONE_UPSTREAM }}#drone-server#g" $file
        sed -i "s#{{ GOGS_UPSTREAM }}#gogs#g" $file
      done
      cd -
    }

    set +e
    test "$ENABLE_NGINX" = 'FALSE' && _sed_external_nginx
    set -e

    test "$ENABLE_NGINX" = 'FALSE' || _sed_nginx

    docker-compose ${COMPOSE_FILE:-} up ${opt:-} ${CI_INCLUDE:-drone-server drone-agent gogs registry} \
       $( test "$ENABLE_NGINX" = 'FALSE' || echo 'nginx' )
}

_tls-config(){
  docker-compose ${COMPOSE_FILE:-} config
}

_down(){
  docker-compose down --remove-orphans
}

_logs(){
  if ! [ -f logs/nginx/access.log ];then
    mkdir -p logs/nginx
    touch logs/nginx/access.log
    touch logs/nginx/error.log
  fi
}

set -e

_logs

if [ "$DEBUG" = 'true' ];then set -x; fi

OS=`uname -s`

test $OS = 'Darwin' && export PATH=/usr/local/opt/gnu-sed/libexec/gnubin:$PATH

_init

. $PWD/.env

. $PWD/secrets/mysql.env

if [ -f $HOME/.bash_profile ];then
  . ~/.bash_profile
fi

test ! -z "$1" || print_help_info

command=$1

shift

if [ -z "$CI_HOST" -a "$command" = 'up' ];then
  exec echo "CI_HOST not set"
fi

if [ -z "$CI_DOMAIN" -a "$command" = 'up-tls' ];then
  exec echo "CI_DOMAIN not set"
fi

for arg in "$@"
do
  test $arg = '--open-port' && COMPOSE_FILE='-f docker-compose.yml -f docker-compose.override.yml -f docker-compose.port.yml'
  test $arg = '-d' && opt='-d'
  [[ $arg =~ --use-external-nginx=* ]] && ENABLE_NGINX=FALSE && NGINX_CONF=$( echo $arg | cut -d '=' -f 2 )
done

_$command "$@"
